util/retry/retry.go

```diff
@@ -14,8 +14,8 @@ import (
 
 // DefaultRetry is a default retry backoff settings when retrying API calls
 var DefaultRetry = wait.Backoff{
```

Comment:
I don't really understand the question. But for us we have patch the executor image with this setup:

- AWS EKS
- AWS VPC CNI
- Calico
- Argo Executor k8sapi
- Heavy loaded Kubernetes Cluster

In our case, randomly, when the wait container try to made a call to the Kubernetes API in order to watch pods state, we receive an i/o timeout answer even if we change the `TimeoutSeconds`in metatv1.ListOptions for the Kubernetes client. In this case the workflow will be set to failed.

In order to be able to use Argo with k8sapi executor we have used the `ExponentialBackoff` function in order to retry the connection in case we don't received any answer from Kubernetes API on the first attempt.

After this implementation even if we don't receive any answer the first time, we have a retry and the workflow never failed.

Reply:
sorry - I should have been clearer - this struct could be used elsewhere in the code - could you please check to see where and if there are other cases - consider the impact of changing the defaults

Reply:
@alexec We face the same issue as well. EKS, calico and istio (I bypass istio for accesses to api-server via an annotation) but I get i/o timeouts in the wait container logs often and the workflow does not make progress

```
time="2020-06-23T16:14:22Z" level=warning msg="Failed to get pod 'for-loop-sd6zv-2540214372': Get https://172.20.0.1:443/api/v1/namespaces/serviceresilience/pods/for-loop-sd6zv-2540214372: dial tcp 172.20.0.1:443: i/o timeout"
time="2020-06-23T16:14:52Z" level=warning msg="Failed to get pod 'for-loop-sd6zv-2540214372': Get https://172.20.0.1:443/api/v1/namespaces/serviceresilience/pods/for-loop-sd6zv-2540214372: dial tcp 172.20.0.1:443: i/o timeout"
time="2020-06-23T16:15:22Z" level=warning msg="Failed to get pod 'for-loop-sd6zv-2540214372': Get https://172.20.0.1:443/api/v1/namespaces/serviceresilience/pods/for-loop-sd6zv-2540214372: dial tcp 172.20.0.1:443: i/o timeout"
```

Reply:
This retry is used elsewhere:

* Getting secrets
* Deleting PDBs
* Getting config map keys

But it seems really rapid, 5 attempts with 10ms wait. @jessesuen do you remember why this is as it is please?

Reply:
5 attempts in 10ms was the original right? https://github.com/argoproj/argo/blob/master/util/retry/retry.go#L18 (master branch)
And also is 5 times in 10 Milliseconds or 5 times with a 10 ms gap between each ?

Reply:
@alexec any thoughts on moving this PR forward?
I consistent get
```
time="2020-07-07T10:13:26Z" level=error msg="executor error: Failed to establish pod watch: Get https://172.20.0.1:443/api/v1/namespaces/serviceresilience/pods?fieldSelector=metadata.name%3Dresilience-workflow-bz4jm-201058130&watch=true: dial tcp 172.20.0.1:443: i/o timeout\ngithub.com/argoproj/argo/errors.Wrap\n\t/go/src/github.com/argoproj/argo/errors/errors.go:88\ngithub.com/argoproj/argo/errors.InternalWrapErrorf\n\t/go/src/github.com/argoproj/argo/errors/errors.go:78\ngithub.com/argoproj/argo/workflow/executor.(*WorkflowExecutor).waitMainContainerStart\n\t/go/src/github.com/argoproj/argo/workflow/executor/executor.go:853\ngithub.com/argoproj/argo/workflow/executor.(*WorkflowExecutor).Wait\n\t/go/src/github.com/argoproj/argo/workflow/executor/executor.go:817\ngithub.com/argoproj/argo/cmd/argoexec/commands.waitContainer\n\t/go/src/github.com/argoproj/argo/cmd/argoexec/commands/wait.go:40\ngithub.com/argoproj/argo/cmd/argoexec/commands.NewWaitCommand.func1\n\t/go/src/github.com/argoproj/argo/cmd/argoexec/commands/wait.go:16\ngithub.com/spf13/cobra.(*Command).execute\n\t/go/pkg/mod/github.com/spf13/cobra@v0.0.4-0.20181021141114-fe5e611709b0/command.go:766\ngithub.com/spf13/cobra.(*Command).ExecuteC\n\t/go/pkg/mod/github.com/spf13/cobra@v0.0.4-0.20181021141114-fe5e611709b0/command.go:852\ngithub.com/spf13/cobra.(*Command).Execute\n\t/go/pkg/mod/github.com/spf13/cobra@v0.0.4-0.20181021141114-fe5e611709b0/command.go:800\nmain.main\n\t/go/src/github.com/argoproj/argo/cmd/argoexec/main.go:17\nruntime.main\n\t/usr/local/go/src/runtime/proc.go:203\nruntime.goexit\n\t/usr/local/go/src/runtime/asm_amd64.s:1357"
time="2020-07-07T10:13:26Z" level=info msg="No output parameters"
time="2020-07-07T10:13:26Z" level=info msg="No output artifacts"
time="2020-07-07T10:13:26Z" level=info msg="No Script output reference in workflow. Capturing script output ignored"
time="2020-07-07T10:13:26Z" level=info msg="Killing sidecars"
time="2020-07-07T10:13:56Z" level=warning msg="Failed to get pod 'resilience-workflow-bz4jm-201058130': Get https://172.20.0.1:443/api/v1/namespaces/serviceresilience/pods/resilience-workflow-bz4jm-201058130: dial tcp 172.20.0.1:443: i/o timeout"
time="2020-07-07T10:14:26Z" level=warning msg="Failed to get pod 'resilience-workflow-bz4jm-201058130': Get https://172.20.0.1:443/api/v1/namespaces/serviceresilience/pods/resilience-workflow-bz4jm-201058130: dial tcp 172.20.0.1:443: i/o timeout"
time="2020-07-07T10:14:56Z" level=warning msg="Failed to get pod 'resilience-workflow-bz4jm-201058130': Get https://172.20.0.1:443/api/v1/namespaces/serviceresilience/pods/resilience-workflow-bz4jm-201058130: dial tcp 172.20.0.1:443: i/o timeout"
time="2020-07-07T10:15:26Z" level=warning msg="Failed to get pod 'resilience-workflow-bz4jm-201058130': Get https://172.20.0.1:443/api/v1/namespaces/serviceresilience/pods/resilience-workflow-bz4jm-201058130: dial tcp 172.20.0.1:443: i/o timeout"
time="2020-07-07T10:15:56Z" level=warning msg="Failed to get pod 'resilience-workflow-bz4jm-201058130': Get https://172.20.0.1:443/api/v1/namespaces/serviceresilience/pods/resilience-workflow-bz4jm-201058130: dial tcp 172.20.0.1:443: i/o timeout"
time="2020-07-07T10:15:56Z" level=error msg="Failed to kill sidecars: Get https://172.20.0.1:443/api/v1/namespaces/serviceresilience/pods/resilience-workflow-bz4jm-201058130: dial tcp 172.20.0.1:443: i/o timeout"
```

The exponential retry in `waitMainContainerStart()` and the increase in the backoff period during retry would both help.

Do you happen to know the request timeout for api-requests made by the executor ?

Reply:
I would create a different retry config for you code tailored to your use case.

Reply:
You can’t use the value used in this PR ?

Reply:
@jsenon is that a question to me ? I dont want to fork and maintain argo-workflows and looking for ways to upstream this and upgrade to the latest version.
@alexec point is, the same retry config is used for multiple things (that might or might not be hitting the api-server) and so using a specific retryConfig for waitContainer related things would be prudent

Reply:
Correct. Lets use a different retry config for this PR rather than change the global one.

Reply:
We experienced a similar issue in a different place

```
time="2020-07-15T13:08:58Z" level=info msg="Starting to wait completion of containerID 3b816ad7e088a81d7c0a21f05a81fcdf44c4e4f9eaf58bc110437e684e5e0186 ..."
time="2020-07-15T13:08:58Z" level=info msg="Starting deadline monitor"
time="2020-07-15T13:12:58Z" level=warning msg="Failed to wait for container id '3b816ad7e088a81d7c0a21f05a81fcdf44c4e4f9eaf58bc110437e684e5e0186': etcdserver: request timed out"
time="2020-07-15T13:12:58Z" level=error msg="executor error: etcdserver: request timed out"
```

It seems triggered by this code: https://github.com/argoproj/argo/blob/master/workflow/executor/executor.go#L918
If the plan is to make a different retry config for this PR, can you also add that for the code above.

The default duration seems a bit small: https://github.com/argoproj/argo/blob/master/util/retry/retry.go#L18

Reply:
So many places need change. Wait, WaitMainContainerStart, KIllSideCar.

KillSideCar timesout 100% of the times when we have more than one side-car (istio-proxy and vault-agent)

Reply:
So, perhaps not a bad idea to change values in the main default retryer

Reply:
Umm.. Default retryer is used for other things though, that @alexec has pointed out. We just need another say DefaultWaitRetry struct to use in all the places we discussed in this thread